### PR TITLE
chore: create redo/undo bridge

### PR DESCRIPTION
**Pull request by qinluhe**
### Feature Preview

---

#### PR Checklist

- [ ] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [ ] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 1]
```diff
@@ -103,6 +104,46 @@ pub fn convert_data_to_document_internal(
...
+
+pub(crate) async fn document_undo(
+  data: AFPluginData<DocumentRedoUndoPayloadPB>,
+  manager: AFPluginState<Arc<DocumentManager>>,
+) -> FlowyResult<String> {
+  let context = data.into_inner();
+  let doc_id = context.document_id;
+  let undo = manager.undo(&doc_id)?;
+  Ok(undo.to_string())
+}
```
Comment (Xazin) [inline frontend/rust-lib/flowy-document2/src/event_handler.rs:125]: Couldn't these be unified into one method?

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 2]
```diff
@@ -103,6 +104,46 @@ pub fn convert_data_to_document_internal(
...
+
+pub(crate) async fn document_can_undo(
+  data: AFPluginData<DocumentRedoUndoPayloadPB>,
+  manager: AFPluginState<Arc<DocumentManager>>,
+) -> FlowyResult<String> {
+  let context = data.into_inner();
+  let doc_id = context.document_id;
+  let can_undo = manager.can_undo(&doc_id)?;
+  Ok(can_undo.to_string())
+}
```
Comment (Xazin) [inline frontend/rust-lib/flowy-document2/src/event_handler.rs:145]: Same here, can also be consolidated.

File: frontend/rust-lib/flowy-document2/src/manager.rs [hunk 3]
```diff
@@ -111,4 +111,42 @@ impl DocumentManager {
...
+  pub fn undo(&self, doc_id: &str) -> FlowyResult<bool> {
+    let uid = self.user.user_id()?;
+    let db = self.user.collab_db()?;
+    let collab = self.collab_builder.build(uid, &doc_id, "document", db);
+    let undo = collab.lock().undo();
+    if let Ok(undo) = undo {
+      return Ok(undo);
+    }
+    Ok(false)
+  }
```
Comment (Xazin) [inline frontend/rust-lib/flowy-document2/src/manager.rs:135]: Can also be consolidated

File: frontend/rust-lib/flowy-document2/src/manager.rs [hunk 4]
```diff
@@ -111,4 +111,42 @@ impl DocumentManager {
...
+    Ok(can_redo)
+  }
+
+  pub fn can_undo(&self, doc_id: &str) -> FlowyResult<bool> {
+    let uid = self.user.user_id()?;
+    let db = self.user.collab_db()?;
+    let collab = self.collab_builder.build(uid, &doc_id, "document", db);
+    let can_undo = collab.lock().can_undo();
+    Ok(can_undo)
+  }
```
Comment (Xazin) [inline frontend/rust-lib/flowy-document2/src/manager.rs:151]: Same

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 5]
```diff
@@ -103,6 +104,62 @@ pub fn convert_data_to_document_internal(
...
+  data: AFPluginData<DocumentRedoUndoPayloadPB>,
+  manager: AFPluginState<Arc<DocumentManager>>,
+) -> DataResult<DocumentRedoUndoResponsePB, FlowyError> {
+  let context = data.into_inner();
+  let doc_id = context.document_id;
+  let document = manager.open_document(doc_id)?;
+  let document = document.lock();
+  let can_redo = document.can_redo();
+  let can_undo = document.can_undo();
+  drop(document);
```
[the remark this answers is not in the document]
Comment (qinluhe) [reply frontend/rust-lib/flowy-document2/src/event_handler.rs:155]: The open_document is get_or_open_document, I  will rename it

**Pull request merged**